airbyte-cdk/python/connector_builder/message_grouper.py

```diff
@@ -0,0 +1,192 @@
+        )
+
+    @staticmethod
+    def _close_page(current_page_request, current_page_response, current_slice_pages, current_page_records):
+        if not current_page_request or not current_page_response:
+            raise ValueError("Every message grouping should have at least one request and response")
+
+        current_slice_pages.append(
+            StreamReadPages(request=current_page_request, response=current_page_response, records=deepcopy(current_page_records))
+        )
+        current_page_records.clear()
+
+    def _create_request_from_log_message(self, log_message: AirbyteLogMessage) -> Optional[HttpRequest]:
+        # TODO: As a temporary stopgap, the CDK emits request data as a log message string. Ideally this should come in the
+        # form of a custom message object defined in the Airbyte protocol, but this unblocks us in the immediate while the
+        # protocol change is worked on.
+        raw_request = log_message.message.partition("request:")[2]
+        try:
+            request = json.loads(raw_request)
+            url = urlparse(request.get("url", ""))
+            full_path = f"{url.scheme}://{url.hostname}{url.path}" if url else ""
+            parameters = parse_qs(url.query) or None
+            return HttpRequest(
+                url=full_path,
+                http_method=request.get("http_method", ""),
+                headers=request.get("headers"),
+                parameters=parameters,
+                body=request.get("body"),
+            )
+        except JSONDecodeError as error:
+            self.logger.warning(f"Failed to parse log message into request object with error: {error}")
+            return None
+
+    def _create_response_from_log_message(self, log_message: AirbyteLogMessage) -> Optional[HttpResponse]:
+        # TODO: As a temporary stopgap, the CDK emits response data as a log message string. Ideally this should come in the
+        # form of a custom message object defined in the Airbyte protocol, but this unblocks us in the immediate while the
+        # protocol change is worked on.
+        raw_response = log_message.message.partition("response:")[2]
+        try:
+            response = json.loads(raw_response)
+            body = response.get("body", "{}")
+            return HttpResponse(status=response.get("status_code"), body=body, headers=response.get("headers"))
+        except JSONDecodeError as error:
+            self.logger.warning(f"Failed to parse log message into response object with error: {error}")
+            return None
+
+    def _has_reached_limit(self, slices):
+        if len(slices) >= self._max_slices:
+            return True
+
+        for slice in slices:
+            if len(slice.pages) >= self._max_pages_per_slice:
+                return True
+        return False
+
+    @classmethod
+    def _create_configure_catalog(cls, stream_name: str) -> ConfiguredAirbyteCatalog:
```

Comment:
creating the catalog on the Java side would allow us to keep the same read interface as connectors